### PR TITLE
Enable casting of string to timestamp with microsecond resolution

### LINKED ISSUE #3751: Support casting string to timestamp with microsecond resolution
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I'm trying to migrate [Seafowl](https://github.com/splitgraph/seafowl)'s storage backend to use [delta-rs](https://github.com/delta-io/delta-rs), and one issue that has come up is the fact that the DELTA protocol [prescribes](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types) that timestamp resolution should be in microseconds. On the other hand, the current casting logic only supports nanoseconds resolution.

**Describe the solution you'd like**
Ideally, we'd be able to cast a `Utf8`/`LargeUtf8` type to `Timestamp(TimeUnit::Microsecond, _)`, like we already do for `Timestamp(TimeUnit::Nanosecond, _)` (even with potential precision loss).

The microsecond resolution of the fractional part of the string timestamp is the default offered for most other systems, e.g. [Postgres](https://www.postgresql.org/docs/current/datatype-datetime.html), [MySQL](https://dev.mysql.com/doc/refman/8.0/en/datetime.html), [BigQuery](https://cloud.google.com/bigquery/docs/reference/standard-sql/data-types#timestamp_type), etc.

**Describe alternatives you've considered**
Perform manual conversion of string to timestamp[us] in some places, possibly prohibiting usage of some high level datafusion/delta-rs interfaces.

Alternatively, enforce users to always call `to_timestamp_micros` datafusion function when specifying timestamp strings.

**Additional context**

### PR DESCRIPTION
**Pull request by gruuya**
# Which issue does this PR close?

Closes #3751.

# Rationale for this change
 
Enable casting strings to timestamps with second, millisecond and microsecond resolution.

# What changes are included in this PR?

Implement conversion of `Utf8`/`LargeUtf8` timestamp strings to `Timestamp(TimeUnit::Second, _)`/`Timestamp(TimeUnit::Millisecond, _)`/`Timestamp(TimeUnit::Microsecond, _)` by making `cast_string_to_timestamp` generic.

# Are there any user-facing changes?

No more casting errors for the said type pairs.

### CODE REVIEW

File: arrow-cast/src/cast.rs [hunk 1]
```diff
@@ -2562,26 +2573,35 @@ fn cast_string_to_timestamp_ns<Offset: OffsetSizeTrait>(
         .downcast_ref::<GenericStringArray<Offset>>()
         .unwrap();
 
+    let scale_factor = match TimestampType::get_time_unit() {
+        TimeUnit::Microsecond => 1_000,
+        TimeUnit::Nanosecond => 1,
+        _ => return Err(ArrowError::CastError("String to timestamp casting supports only microsecond or nanosecond resolution".to_string()))
```
Comment (tustvold) [inline arrow-cast/src/cast.rs:2579]: Is there a particular reason not to just support the full set whilst here?

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
I have confirmed that the integration test in array_cast already contains examples of other timestamp array types

File: arrow-cast/src/cast.rs (hunk 1 above)
Comment (gruuya) [reply arrow-cast/src/cast.rs:2579]: Not to my knowledge; I can add them too if no one objects.

**Review by tustvold — APPROVED** [review]

**Comment by tustvold** [comment]
Thank you

**Pull request merged**